모임 생성 기능 로직 작성

qaws5656 opened the pull request:
<!-- PULL REQUEST TEMPLATE -->
<!-- (체크박스 "[ ]"를 "[x]"로 작성하여, 체크해주세요) -->

## 해당 사항 (중복 선택)

<!-- 해당되는 사항만 남기고 나머지 줄을 삭제해주세요 -->

- [X] FEAT : 새로운 기능 추가 및 개선

## 설명

### 이슈 번호

 resolve #4 

<!-- 키워드를 사용해 이슈를 연결해주세요 -->
<!-- 예시: close #1 / closes #1, #3 / resolve #4 -->

### Key Changes

- 모임 생성 API에 필요한 로직 작성 및 테스트 코드 작성
- GatheringRegisterService -> 모임 생성을 위한 비즈니스 로직을 처리합니다.
- GatheringMemberAppender -> 멤버 추가 로직을 실질적으로 구현하는 클래스입니다. 멤버 추가에 필요한 유효성 검증을 수행하고 멤버 추가 처리를 합니다.

### How it Works

- 모임 생성 시 모임 정보 데이터를 저장할 뿐만 아니라 모임을 생성 한 방장을 Member로 추가해야하기에 모임 생성자를 Member로 
추가하는 로직도 수행됩니다.

#### Work list
 - [x] 모임 생성 후 DB에 저장하는 로직을 작성합니다.
 - [x] 모임 구성원을 추가하고 DB에 저장하는 로직을 작성합니다.

### To Reviewers

- 기존에는 core 모듈에 GatheringRegister 클래스를 생성하고 register 메서드를 정의하여 repository.save를 처리하였습니다.
이 후 api 모듈의 GatheringRegisterService에서 주입 받아 DB 저장 로직을 처리하였는데요.
 아직까지는 모임 등록 과정에서 복잡한 요구 사항이 없어서 불필요한 계층이 하나 더 생긴 것 같아
 GatheringRegisterService에서 gatheringRepository.save 를 직접 호출하는 방식으로 수정하였습니다.
추후에 모임 등록 과정에서 복잡한 요구 사항이 생기면 그 때 GatheringRegister 클래스를 만들어 처리할 지
지금부터라도 단순한 CRUD일지라도 따로 core 모듈에 클래스를 만들어 처리할 지 고민됩니다.

- 추가로 제가 테스트 코드를 최근에 강의로 배우며 시작하고 있는데 Classicist VS. Mockist 관련해서 진영 싸움이 있다고 들었습니다 
ㅎㅎ..
저는 일단 강의에서 Classicist 방식으로 배우고 있어서 작성한 테스트 코드에 SpringBootTest 어노테이션을 사용하면서 통합 테스트 방식으로 작성하였는데 어떠신지 궁금합니다!! 하하

qaws5656 committed:
feat: Gathering 도메인 Repository 추가

qaws5656 committed:
feat: GatheringRepository save 메서드 추가

qaws5656 committed:
feat: GatheringRegister.register 메서드 추가

qaws5656 committed:
feat: GatheringMember 도메인 Repository 추가

qaws5656 committed:
feat: GatheringMemberRepository save 메서드, isExistsByUserIdAndGatheringId 메서드 추가

qaws5656 committed:
feat: Gathering 도메인 participantCount 필드 추가
 - 현재 멤버 정원을 의미

qaws5656 committed:
feat: 모임 구성원 추가 관련 로직 작성
 - 유효한 모임인지, 모집 중인 모임인지, 인원이 다 찼는지, 이미 가입 한 모임인지 검사
 - GatheringMemberAppender 클래스에서 처리

qaws5656 committed:
feat: 모임 생성 API 요청을 처리하는 GatheringRegisterService 추가
 - external-api 모듈에 위치하며 필요한 비즈니스 로직을 수행

qaws5656 committed:
fix: isRecurring -> isRecruiting 메서드 이름 오타 수정

qaws5656 committed:
test: Gathering 도메인 테스트 코드 작성

qaws5656 committed:
test: GatheringRepository 테스트 코드 작성

qaws5656 committed:
refactor: 모임을 등록하는 로직은 현재 단순 저장 로직이므로 GatheringRegister가 아닌 Repository를 직접 이용하도록 변경
 - 추후 모임을 등록하는 로직이 복잡해지면 다시 하위 구현체 (GatheringRegister)를 이용하여 처리

qaws5656 committed:
test: GatheringRegisterService.register 코드 작성

qaws5656 committed:
feat: GatheringRegisterService.register 트랜젝션 처리

qaws5656 committed:
test GatheringMemberAppender 테스트 코드 작성

qaws5656 committed:
test GatheringMemberRepository 테스트 코드 작성

qaws5656 committed:
refactor: GatheringMemberRepository 메서드 이름 변경
 - isExistsByUserIdAndGatheringId -> isGatheringMember

oereo commented on momoim-core/src/test/java/com/triplem/momoim/core/domain/gathering/GatheringRepositoryTest.java:
```diff
@@ -0,0 +1,92 @@
+    void saveGathering() {
+        //given
+        LocalDateTime now = LocalDateTime.now();
+        Gathering gathering = Gathering.builder()
```
Gathering 객체를 만드는 행위를 fixture로 분리하거나! setup method를 통해서 사용해도 괜찮을 것 같습니다!! saveGathering method와 findGatheringById method에서 같은 객체를 생성하는 부분이 보여서요!!

qaws5656 replied:
아 저도 테스트 코드를 작성하면서 중복되는 테스트용 객체를 클래스 하나에 
그냥 메소드처럼 만들어둘까 했는데 fixture라는 용어가 있었군요.
감사합니다!

oereo replied:
아닙니다!! 쿄쿄쿄

oereo commented on momoim-core/src/test/java/com/triplem/momoim/core/domain/member/GatheringMemberRepositoryTest.java:
```diff
@@ -0,0 +1,63 @@
+    @DisplayName("모임 멤버를 저장할 수 있다.")
+    void saveGatheringMember() {
+        //given
+        GatheringMember gatheringMember = GatheringMember.create(1L, 5L);
```
이거 보다가 갑자기 드는 생각인데요! 처음에 `1L`과 `5L`이 의미하는게 뭔지 헷갈릴수가 있을 것 같네요 ㅠㅠㅠ
builder 패턴으로 생성하는게 나을지 고민이 약간 들지만! 로직에 문제가 있지 않아서! 수정을 따로 하시지는 않아도 될 것 같습니다!

qaws5656 replied:
혹시 테스트 코드에서 헷갈린다는 뜻이신가요??
아니면 create 메서드 자체가 헷갈리신다는 건가요? 최근에 저도 일반 생성자, 빌더 패턴에서 
팩토리 메서드를 사용하는 방식으로 시작해서 여쭤봅니다!
빌더 패턴으로 수정하는 것이 더 가독성이 좋다고 느끼시면 저는 상관없습니다!

oereo replied:
아 테스트 코드에서 헷갈리는 것보다 create method를 사용할 때 들어가는 parameter의 의미 전달성에 대한 부분이었습니다!!! 같은 long type의 데이터가 인자로 들어가다보니 가독성과 의미전달이 어떤 방식이 좋을지 고민되네요!

qaws5656 replied:
실제 Serivce Layer에서 사용될 때는 
GatheringMember.create(userId, gatheringId) 이런식으로 사용되서 구분은 되지 않을까 싶습니다

oereo replied:
아합! 좋습니다!!! 팩토리로 가시죠!

oereo commented on momoim-external-api/src/main/java/com/triplem/momoim/api/gathering/GatheringRegisterService.java:
```diff
@@ -0,0 +1,22 @@
+@Service
+@RequiredArgsConstructor
+public class GatheringRegisterService {
+    private final GatheringRepository gatheringRepository;
```
이 부분이 약간 고민이 되는군요 ㅠㅠ
Layer의 depth를 좀 맞추고 싶은 생각이 있긴하네요 ㅠ
예를들어 GatheringRepository, GatheringMemberRepository를 주입받는 방식과 GatheringRegister와 같은 layer를 하나 더 만들어서 사용하는 방법중에 고민이긴한데요!

GatheringMemberAppender에서 validation을 수행이 된다고 하면, GatheringRegister와 같은 layer에 대한 class를 정의하고 Gathering을 생성할 때에 대한 validation 로직들을 넣는 것도 맥락을 맞출 때 좋을 것 같아요!

불필요한 Layer 추가된다고 생각이 드시면 아예 repository layer에 대해서 의존성을 주입받는 식으로 진행하셔도 될 것 같습니다!

qaws5656 replied:
계층 구조에 대해서 한번 의논해봐야겠네요!

oereo replied:
좋습니다!!!

oereo reviewed (commented):
너무 고생많으셨습니다!! 리뷰 3건 있긴한데요! 한번 확인 부탁드릴게요!!

oereo commented:
```
기존에는 core 모듈에 GatheringRegister 클래스를 생성하고 register 메서드를 정의하여 repository.save를 처리하였습니다.
이 후 api 모듈의 GatheringRegisterService에서 주입 받아 DB 저장 로직을 처리하였는데요.
아직까지는 모임 등록 과정에서 복잡한 요구 사항이 없어서 불필요한 계층이 하나 더 생긴 것 같아
GatheringRegisterService에서 gatheringRepository.save 를 직접 호출하는 방식으로 수정하였습니다.
추후에 모임 등록 과정에서 복잡한 요구 사항이 생기면 그 때 GatheringRegister 클래스를 만들어 처리할 지
지금부터라도 단순한 CRUD일지라도 따로 core 모듈에 클래스를 만들어 처리할 지 고민됩니다.

추가로 제가 테스트 코드를 최근에 강의로 배우며 시작하고 있는데 Classicist VS. Mockist 관련해서 진영 싸움이 있다고 들었습니다
ㅎㅎ..
저는 일단 강의에서 Classicist 방식으로 배우고 있어서 작성한 테스트 코드에 SpringBootTest 어노테이션을 사용하면서 통합 테스트 방식으로 작성하였는데 어떠신지 궁금합니다!! 하하
```

좋은 주제네요!! 저는 그냥 GatheringRegister 라는 클래스를 만드는 것도 좋을 것 같아요!! 저도 이와 비슷한 형식으로 로직을 구성했었던 적이 있어서요!!

Classicist 방식에 대해서 저도 한번 보고 승환님의 방향으로 따라가겠습니다!! 통합 테스트 방식 좋긴한데요! 저는 unit test를 같이 넣어서 작업하는것도 좋을 것 같습니다!! 예를들어 여러 layer에 대해서 실제 test db를 통해서 테스트가 필요하면 통합테스트로 하는것이 맞는데, 간단한 객체에 대한 테스트의 경우에는 unit test 방식으로 구성하는게 속도면에서 좋을 것 같습니다!!

qaws5656 committed:
test: GatheringFixture 클래스 추가
 - 테스트를 위한 Gathering 객체를 생성하는 객체로 managerId, recruitStatus, capacity, participantCount는 테스트 조건에 사용되는 데이터이므로 임의 값으로 지정하여 생성할 수 있도록 함

qaws5656 committed:
test: GatheringTest 테스트 코드 GatheringFixture 사용하는 방식으로 변경

qaws5656 committed:
test: GatheringRepositoryTest 테스트 코드 GatheringFixture 사용하는 방식으로 변경

qaws5656 committed:
test: GatheringMemberAppenderTest 테스트 코드 GatheringFixture 사용하는 방식으로 변경

qaws5656 committed:
test: GatheringMemberAppenderTest 테스트 코드 추가
 - 존재하지 않는 모임에 멤버 추가가 될 수 없다.

oereo commented:
승환님! 반영되시는대로 머지하셔도 될 것 같습니다!! 고생많으셨습니다!!
